cookie_source.go

```diff
@@ -11,63 +11,68 @@ import (
 	"golang.org/x/crypto/hkdf"
 )
 
-// CookieSource is used to create and verify source address tokens
+// CookieSource is used to create and verify a cookie
 type CookieSource interface {
 	// NewToken creates a new token
 	NewToken([]byte) ([]byte, error)
 	// DecodeToken decodes a token
 	DecodeToken([]byte) ([]byte, error)
 }
 
-type defaultCookieSource struct {
-	aead cipher.AEAD
+const cookieSecretSize = 32
+const cookieNonceSize = 32
+
+// The DefaultCookieSource is a simple implementation for the CookieSource.
+type DefaultCookieSource struct {
+	secret []byte
 }
 
-const tokenKeySize = 16
-const tokenNonceSize = 16
+var _ CookieSource = &DefaultCookieSource{}
 
-// newDefaultCookieSource creates a source for source address tokens
-func newDefaultCookieSource() (CookieSource, error) {
+// NewDefaultCookieSource creates a source for source address tokens
+func NewDefaultCookieSource() (CookieSource, error) {
 	secret := make([]byte, 32)
 	if _, err := rand.Read(secret); err != nil {
 		return nil, err
 	}
-	key, err := deriveKey(secret)
-	if err != nil {
-		return nil, err
-	}
-	c, err := aes.NewCipher(key)
-	if err != nil {
-		return nil, err
-	}
-	aead, err := cipher.NewGCMWithNonceSize(c, tokenNonceSize)
-	if err != nil {
-		return nil, err
-	}
-	return &defaultCookieSource{aead: aead}, nil
+	return &DefaultCookieSource{secret: secret}, nil
 }
 
-func (s *defaultCookieSource) NewToken(data []byte) ([]byte, error) {
-	nonce := make([]byte, tokenNonceSize)
+// NewToken encodes data into a new token.
+func (s *DefaultCookieSource) NewToken(data []byte) ([]byte, error) {
+	nonce := make([]byte, cookieNonceSize)
 	if _, err := rand.Read(nonce); err != nil {
 		return nil, err
 	}
-	return s.aead.Seal(nonce, nonce, data, nil), nil
+	aead, err := s.createAEAD(nonce)
+	if err != nil {
+		return nil, err
+	}
+	return aead.Seal(nonce, nonce, data, nil), nil
```

Comment:
I don't think this is quite what you want. Rather, you want to do:

```
r := make([]byte, 32)
rand.Read(r)
key = hkdf(secret, R, []byte("mint cookie key")) // key length
nonce = hkdf(secret, R, []byte("mint cookie nonce")) // 12 bytes
```

And then you can use nonce the way you have it here.

Reply:
Or, if you prefer you can just read the key and the nonce as successive values from the hkdf.

Reply:
How is that supposed to work? When decoding the cookie, I don't know `r` any more, and I can't create the AEAD to open the cookie.

Reply:
The cookie is r || aead.Seal(blah...)

Right now you are just using nonce == r, so the difference I am proposing is you hkdf r to make the nonce.